### DISCUSSION TITLE
Outputting/Evaluating fields at the bottom of an immersed boundary grid

### DISCUSSION THREAD

**Discussion by francispoulin** [CliMA/Oceananigans.jl#3032] [General]
Sorry that I don't know this but when using an immersed boundary method, what's the best way to evaluate a field at the bottom, right at the bathymetry?

**Comment by simone-silvestri** [comment]
I am not sure I understood correctly, but do you want a 2D field that in `i, j` has the value that (in 3D) corresponds to the cell just above bathymetry? If this is the case, we don't have a ready-to-go functionality, but you can use conditional operations

The cell just above bathymetry is the only cell in the column for which `immersed_peripheral_node(i, j, k-1, grid, LX, LY, LZ) == true` (note this holds only for 2D immersed map, not for a 3D mask)

So you can do

```julia
julia> grid = RectilinearGrid(size = (10, 10, 10), extent = (1, 1, 1))
10×10×10 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── Periodic x ∈ [0.0, 1.0)  regularly spaced with Δx=0.1
├── Periodic y ∈ [0.0, 1.0)  regularly spaced with Δy=0.1
└── Bounded  z ∈ [-1.0, 0.0] regularly spaced with Δz=0.1

julia> ibg = ImmersedBoundaryGrid(grid, GridFittedBottom((x, y) -> - x))
10×10×10 ImmersedBoundaryGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo:
├── immersed_boundary: GridFittedBottom(min(h)=-9.50e-01, max(h)=-5.00e-02)
├── underlying_grid: 10×10×10 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── Periodic x ∈ [0.0, 1.0)  regularly spaced with Δx=0.1
├── Periodic y ∈ [0.0, 1.0)  regularly spaced with Δy=0.1
└── Bounded  z ∈ [-1.0, 0.0] regularly spaced with Δz=0.1

julia> field = CenterField(ibg)
10×10×10 Field{Center, Center, Center} on ImmersedBoundaryGrid on CPU
├── grid: 10×10×10 ImmersedBoundaryGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 16×16×16 OffsetArray(::Array{Float64, 3}, -2:13, -2:13, -2:13) with eltype Float64 with indices -2:13×-2:13×-2:13
    └── max=0.0, min=0.0, mean=0.0

julia> set!(field, (x, y, z) -> x)
10×10×10 Field{Center, Center, Center} on ImmersedBoundaryGrid on CPU
├── grid: 10×10×10 ImmersedBoundaryGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 16×16×16 OffsetArray(::Array{Float64, 3}, -2:13, -2:13, -2:13) with eltype Float64 with indices -2:13×-2:13×-2:13
    └── max=0.95, min=0.15, mean=0.683333

julia> using Oceananigans.ImmersedBoundaries: immersed_peripheral_node

julia> @inline bottom_condition(i, j, k, grid) = immersed_peripheral_node(i, j, k-1, grid, Center(), Center(), Center())
bottom_condition (generic function with 1 method)

julia> using Oceananigans.Fields: condition_operand

julia> bottom_height = sum(condition_operand(identity, field, bottom_condition, 0.0), dims = 3)
10×10×1 Field{Center, Center, Nothing} reduced over dims = (3,) on ImmersedBoundaryGrid on CPU
├── grid: 10×10×10 ImmersedBoundaryGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: Nothing, top: Nothing, immersed: ZeroFlux
└── data: 16×16×1 OffsetArray(::Array{Float64, 3}, -2:13, -2:13, 1:1) with eltype Float64 with indices -2:13×-2:13×1:1
    └── max=0.95, min=0.15, mean=0.55

julia> interior(bottom_height)
10×10×1 view(::Array{Float64, 3}, 4:13, 4:13, 1:1) with eltype Float64:
[:, :, 1] =
 0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0
 0.15  0.15  0.15  0.15  0.15  0.15  0.15  0.15  0.15  0.15
 0.25  0.25  0.25  0.25  0.25  0.25  0.25  0.25  0.25  0.25
 0.35  0.35  0.35  0.35  0.35  0.35  0.35  0.35  0.35  0.35
 0.45  0.45  0.45  0.45  0.45  0.45  0.45  0.45  0.45  0.45
 0.55  0.55  0.55  0.55  0.55  0.55  0.55  0.55  0.55  0.55
 0.65  0.65  0.65  0.65  0.65  0.65  0.65  0.65  0.65  0.65
 0.75  0.75  0.75  0.75  0.75  0.75  0.75  0.75  0.75  0.75
 0.85  0.85  0.85  0.85  0.85  0.85  0.85  0.85  0.85  0.85
 0.95  0.95  0.95  0.95  0.95  0.95  0.95  0.95  0.95  0.95

```

**Reply by francispoulin** [reply]
That is exactly what I was trying to ask.  Sorry that I was unclear but thanks very much for the answer!  I will try it out and let you know how it goes.

**Reply by francispoulin** [reply]
I have tried what was suggested but have not figured it out.   Sorry.  Below is what I have so far.

Note that at the moment I am writing the pressure over the volume, but I first want to plot the pressure at the bottom, and then I want to plot an `x` derivative of the pressure.

I know that I need to change the `set!` command and then need to specify the output in the output writer, but I have not figured out the correct syntax.

Can you help me figure out what I am doing wrong?

```
using Oceananigans
using Oceananigans.ImmersedBoundaries: ImmersedBoundaryGrid, GridFittedBottom, immersed_peripheral_node
using Oceananigans.Fields: condition_operand

underlying_grid = RectilinearGrid(; size = (6, 6, 6), halo = (4, 4, 4),
                                       x = (0, 1), y = (0, 1), z = (-1, 0),
                                topology = (Bounded, Bounded, Bounded))

grid = ImmersedBoundaryGrid(underlying_grid, GridFittedBottom((x, y) -> - x))
model = HydrostaticFreeSurfaceModel(grid = grid, free_surface = ImplicitFreeSurface())

simulation = Simulation(model, Δt=0.1, stop_time=0.5)

@inline bottom_condition(i, j, k, grid) = immersed_peripheral_node(i, j, k-1, grid, Center(), Center(), Center())

pressure = CenterField(grid)
# FJP: Want to save the bottom pressure: model.pressure.pHY′ on bathymetry
set!(pressure, (x, y, z) -> x)

bottom_pressure = sum(condition_operand(identity, pressure, bottom_condition, 0.0), dims = 3)
field_to_save = interior(bottom_pressure)

#FJP: want to output the bottom pressure in the writer
simulation.output_writers[:fields] = JLD2OutputWriter(model, (model.pressure.pHY′,);
                                                      filename = "p_bottom_field",
                                                      schedule = TimeInterval(0.1),
                                                      overwrite_existing = true,
                                                      with_halos= true)

run!(simulation)
```

**Reply by simone-silvestri** [reply]
If you want to save it you have to do something like this.
```
bottom_field = Field(Average(pressure, condition = bottom_condition, mask = 0.0, dims = 3))
```
and then save bottom_field. I am not 100% sure that it will work. conditional `Reductions` are not thoroughly tested

However, bottom_field will be defined at different heights so I am not sure how to best take an x-derivative.
The order of operations actually matters here, i.e., taking a derivative of bottom_field will return you the difference in x of bottom pressure at different heights divided by the spacing (so not really an x-derivative), while deriving pressure then conditioning will give you the x-derivative of pressure at the bottom.

**Reply by francispoulin** [reply]
Thank you @simone-silvestri !  That is very helpful and I think I have it working.  My code is copied below.

I had a typo before.  Since the topography differs in the `x` direction we need to compute the `y` derivative.  But it makes more sense to me to project the derivative of the pressure onto the top of the bathymetry.  That is what I am trying to do now but if something looks wrong, please let me know.

This is part of a process to compute the bottom pressure torque in a wind-driven gyre case. In general, we will need to compute `J(p_bottom, H)`, so we will need to compute both x and y derivatives of pressure.  

When I get this working, we could try applying this to the global simulations if you are interested?   Yet another validation I suppose but you can't have too many.

```
using Oceananigans
using Oceananigans.ImmersedBoundaries: ImmersedBoundaryGrid, GridFittedBottom, immersed_peripheral_node
using Oceananigans.Fields: condition_operand

underlying_grid = RectilinearGrid(; size = (6, 6, 6), halo = (4, 4, 4),
                                       x = (0, 1), y = (0, 1), z = (-1, 0),
                                topology = (Bounded, Bounded, Bounded))

grid = ImmersedBoundaryGrid(underlying_grid, GridFittedBottom((x, y) -> - x))
model = HydrostaticFreeSurfaceModel(grid = grid, free_surface = ImplicitFreeSurface())

simulation = Simulation(model, Δt=0.1, stop_time=0.5)

dpressure = CenterField(grid)
set!(dpressure, ∂y(model.pressure.pHY′))
@inline bottom_condition(i, j, k, grid) = immersed_peripheral_node(i, j, k-1, grid, Center(), Center(), Center())
bottom_field = Field(Average(dpressure, condition = bottom_condition, mask = 0.0, dims = 3))

simulation.output_writers[:fields] = JLD2OutputWriter(model, (bottom_field,);
                                                      filename = "p_bottom_field",
                                                      schedule = TimeInterval(0.1),
                                                      overwrite_existing = true,
                                                      with_halos= true)
run!(simulation)

```

**Reply by simone-silvestri** [reply]
There is a small problem with `set!(dpressure, ∂y(model.pressure.pHY′))`. This will just replace the data in `dpressure` with the data calculated from `∂y(model.pressure.pHY′)` at the time you call it. 
You want to set up the operation, not the data, so you should do `dpressure =  ∂y(model.pressure.pHY′)`.
For the x-derivative then you can probably calculate it before conditional averaging.

Let me know how it goes, it would be nice to apply it to the global simulations!

**Reply by francispoulin** [reply]
Thanks and I will keep you posted.  Likely sometime next week I will get back to you with working code.

**Reply by navidcy** [reply]
@francispoulin did this work for you? @djlikesdjs was asking me how to do precisely this.

**Reply by francispoulin** [reply]
It did.  I can share sample code a bit later if you want?

**Reply by navidcy** [reply]
Yes please. That’d be sweet.

**Reply by francispoulin** [reply]
Will be here in 2 hours, after I get to a computer.

**Reply by francispoulin** [reply]
I have not tried this for a while so I can't say it will work for sure but I believe this worked for me.
If you find it doesn't work I can look into it tomorrow.

```dpressure = CenterField(grid)
dpressure =  ∂y(model.pressure.pHY′)
@inline bottom_condition(i, j, k, grid) = immersed_peripheral_node(i, j, k-1, grid, Center(), Center(), Center())
bottom_field = Field(Average(dpressure, condition = bottom_condition, mask = 0.0, dims = 3))
```

**Reply by navidcy** [reply]
So above it seems that `dpressure` lives in (Center, Face, Center) (since it's the y-derivative of a field that lives on (Center, Center, Center)) while `bottom_condition` is evaluated at (Center, Center, Center), right? Is this an oversight or on purpose?

**Reply by navidcy** [reply]
And, to make sure I understand, the idea is to "average" in 3rd dimension with the mask that is everywhere 0 except 1 at the bottom-most cell, right?

**Reply by simone-silvestri** [reply]
Yeah, that is the idea. You're right, the condition should be evaluated at `(Center, Face, Center)`

**Reply by francispoulin** [reply]
Yes, we did get the position wrong.  Sorry about that and thanks for pointing that out.

Also yes, we were using this with the immersed boundary method and we were computing the 2D streamfunction  by integrating w.r.t. the third coordinate, and why we used masking.

A simpler example would have been much better.  Maybe you can post yours when you have it working?

**Reply by glwagner** [reply]
Doesn't using `sum` require evaluations at every vertical index? I think a better solution would involve precomputing the vertical index `k` associated with the bottommost cell.

**Reply by francispoulin** [reply]
I agree that there is room for improvement here, and I appreciate the suggestion.  

We integrate with respect to height because we need to integrate the equation for the streamfunction, something like

u = - psi_z

Since they velocity is zero in the boundary region then I thought that summing over the entire depth would be correct and easy.

Maybe I am missing something?

**Reply by glwagner** [reply]
@simone-silvestri's suggestion is to perform a `sum` over an expression which is 0 everywhere except in the cell just above the bottom. This isolates the value of the expression in the cell above the bottom.

There are two problems with this approach. The first is that it is difficult to understand. I understand it because I understand conditional operations, but I am not sure the average user grasps what is going on under the hood immediately. It's confusing that `sum` somehow isolates the value of an expression just above the bottom. The other issue is that it is inefficient. For example, if the expression we want to evaluate is very expensive, the resulting diagnostic will be extremely expensive because the expression is reduced over a column. Worse, it may fail to compile because expressions within reductions fail at lower complexity.

Avoiding these limitations is straightforward in principle since we simply need to precompute the vertical index of the cell above the bottom. However, it will require a new feature in the source code. I think the right abstraction is a wrapper around `AbstractOperation` (not so different from `ConditionalOperation`) that allows one index to be computed / accessed as a function of the other indices at it's `Nothing` location. Then we'd have something like

```julia
@inline Base.getindex(op::ComputedIndexOperation{Center, Center, Nothing}, i, j, k) = op[i, j, op.index[i, j])
```

**Reply by francispoulin** [reply]
Your points are well taken.  I am all for clarity and efficiency.  

I feel like I have seen a function that finds the index right above the topography before, or maybe even wrote it once?  I think it was brute force, which is not very expensive but since you only do it once at the beginning, why not?  

I like this idea.  Thanks.

**Reply by glwagner** [reply]
Above, @simone-silvestri says that

```julia
immersed_peripheral_node(i, j, k-1, grid, LX, LY, LZ) == true
```

implies that `i, j, k` is right above topography, so we can use that.

**Reply by navidcy** [reply]
Just to make a note, whether is the `sum` with conditional operations or anything else we settle on, we should add some user functionality so that users can output bottom fields without need to know the algorithm behind it? So let's find the most efficient way to do it!

(I agree it took me a bit to understand what the conditional sum was achieving here.)

Why don't we compute the map with the indices that correspond to the bottom values can be computed when we construct the immersed boundary grid and then be part of the IBG properties? Then this would be part of all fields that live on the IBG. Wouldn't that be helpful?

**Reply by francispoulin** [reply]
I think having a function that outputs the fields along the immersed boundary is a great idea.  Then a second function, that can sum over the physical domain, would also be very helpful.

I am happy to work on this with someone else if there is interest.

**Reply by glwagner** [reply]
Generally our philosophy is only to compute things as they are needed and store information locally. So I'm not sure we want to precompute the index map "just in case" a user wants it. The other problem is where to put it. `GridFittedBottom` is in principle agnostic to the grid (ie the bottom height could be a function of `x, y`). On the other hand `ImmersedBoundaryGrid` is more general and can be used for 3D immersed boundaries (holes, caves, bridges), not just "2.5D" bathymetry.

I like the name `MappedIndexOperation`

**Comment by navidcy** [comment]
OK, fair point. We do not need to precompute anything. Just create the functionality for the user to do it just by prescribing a kwarg or something.
